File: apps/demo-be-csas/tests/DemoBeCsas.Tests/Integration/Steps/AdminSteps.cs
using System.Text.Json;
using DemoBeCsas.Tests.ScenarioContext;
using FluentAssertions;
using Reqnroll;
using Xunit;

namespace DemoBeCsas.Tests.Integration.Steps;

[Binding]
[Trait("Category", "Integration")]
public class AdminSteps(SharedState state, AuthSteps auth)
{
    // ─────────────────────────────────────────────────────────────
    // When steps
    // ─────────────────────────────────────────────────────────────

    [When(@"^the admin sends GET /api/v1/admin/users$")]
    public async Task WhenAdminListsUsers()
    {
        var client = auth.AdminClient();
        state.LastResponse = await client.GetAsync("/api/v1/admin/users?page=1&size=20");
    }

    [When(@"^the admin sends GET /api/v1/admin/users\?email=alice@example\.com$")]
    public async Task WhenAdminSearchesByEmail()
    {
        var client = auth.AdminClient();
        state.LastResponse = await client.GetAsync(
            "/api/v1/admin/users?email=alice@example.com&page=1&size=20"
        );
    }

    [When(@"^the admin sends POST /api/v1/admin/users/\{alice_id\}/disable with body \{ ""reason"": ""([^""]+)"" \}$")]
    public async Task WhenAdminDisablesAlice(string reason)
    {
        var aliceId = state.LastCreatedId ?? auth._aliceId;
        aliceId.Should().NotBeNull("alice's ID should be known");
        var client = auth.AdminClient();
        state.LastResponse = await client.PostAsync($"/api/v1/admin/users/{aliceId}/disable", null);
    }

    [When(@"^the admin sends POST /api/v1/admin/users/\{alice_id\}/enable$")]
    public async Task WhenAdminEnablesAlice()
    {
        var aliceId = state.LastCreatedId ?? auth._aliceId;
        aliceId.Should().NotBeNull("alice's ID should be known");
        var client = auth.AdminClient();
        state.LastResponse = await client.PostAsync($"/api/v1/admin/users/{aliceId}/enable", null);
    }

    [When(@"^the admin sends POST /api/v1/admin/users/\{alice_id\}/unlock$")]
    public async Task WhenAdminUnlocksAlice()
    {
        var aliceId = state.LastCreatedId ?? auth._aliceId;
        aliceId.Should().NotBeNull("alice's ID should be known");
        var client = auth.AdminClient();
        state.LastResponse = await client.PostAsync($"/api/v1/admin/users/{aliceId}/unlock", null);
    }

    [When(@"^the admin sends POST /api/v1/admin/users/\{alice_id\}/force-password-reset$")]
    public async Task WhenAdminForcePasswordReset()
    {
        var aliceId = state.LastCreatedId ?? auth._aliceId;
        aliceId.Should().NotBeNull("alice's ID should be known");
        var client = auth.AdminClient();
        state.LastResponse = await client.PostAsync(
            $"/api/v1/admin/users/{aliceId}/force-password-reset",
            null
        );
        if (state.LastResponse.IsSuccessStatusCode)
        {
            var body = await state.LastResponse.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(body);
            if (doc.RootElement.TryGetProperty("reset_token", out var rt))
            {
                state.LastResetToken = rt.GetString();
            }
        }
    }

    // ─────────────────────────────────────────────────────────────
    // Then steps
    // ─────────────────────────────────────────────────────────────

    [Then(@"^the response body should contain at least one user with ""email"" equal to ""([^""]+)""$")]
    public async Task ThenResponseContainsUserWithEmail(string email)
    {
        state.LastResponse.Should().NotBeNull();
        var body = await state.LastResponse!.Content.ReadAsStringAsync();
        body.Should().Contain(email, $"Expected user with email '{email}' in: {body}");
    }
}

File: apps/demo-be-java-vertx/src/test/java/com/organiclever/demojavx/integration/steps/AdminSteps.java
package com.organiclever.demojavx.integration.steps;

import com.organiclever.demojavx.support.AppFactory;
import com.organiclever.demojavx.support.ScenarioState;
import com.organiclever.demojavx.support.ServiceResponse;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.junit.jupiter.api.Assertions;

public class AdminSteps {

    private final ScenarioState state;

    public AdminSteps(ScenarioState state) {
        this.state = state;
    }

    @Given("users {string}, {string}, and {string} are registered")
    public void usersAreRegistered(String u1, String u2, String u3) throws Exception {
        AuthSteps authSteps = new AuthSteps(state);
        authSteps.registerUser(u1, u1 + "@example.com", "Str0ng#Pass1");
        authSteps.registerUser(u2, u2 + "@example.com", "Str0ng#Pass1");
        authSteps.registerUser(u3, u3 + "@example.com", "Str0ng#Pass1");
        ServiceResponse listResp = AppFactory.getService()
                .adminListUsers(state.getAdminAccessToken(), null, 1, 100);
        JsonObject body = listResp.body();
        Assertions.assertNotNull(body);
        JsonArray data = body.getJsonArray("data");
        String aliceId = SecuritySteps.findUserIdByUsername(data, "alice");
        state.setUserId(aliceId);
    }

    @When("^the admin sends GET /api/v1/admin/users$")
    public void adminSendsGetUsers() throws Exception {
        ServiceResponse response = AppFactory.getService()
                .adminListUsers(state.getAdminAccessToken(), null, 1, 20);
        state.setLastResponse(response);
    }

    @When("^the admin sends GET /api/v1/admin/users\\?email=(.+)$")
    public void adminSendsGetUsersWithEmailFilter(String email) throws Exception {
        ServiceResponse response = AppFactory.getService()
                .adminListUsers(state.getAdminAccessToken(), email, 1, 20);
        state.setLastResponse(response);
    }

    @When("^the admin sends POST /api/v1/admin/users/\\{alice_id\\}/disable with body \\{ \"reason\": \"([^\"]+)\" \\}$")
    public void adminSendsDisableUser(String reason) throws Exception {
        String userId = state.getUserId();
        Assertions.assertNotNull(userId, "Alice's user ID must be set");
        ServiceResponse response = AppFactory.getService()
                .adminDisableUser(state.getAdminAccessToken(), userId);
        state.setLastResponse(response);
    }

    @When("^the admin sends POST /api/v1/admin/users/\\{alice_id\\}/enable$")
    public void adminSendsEnableUser() throws Exception {
        String userId = state.getUserId();
        Assertions.assertNotNull(userId, "Alice's user ID must be set");
        ServiceResponse response = AppFactory.getService()
                .adminEnableUser(state.getAdminAccessToken(), userId);
        state.setLastResponse(response);
    }

    @When("^the admin sends POST /api/v1/admin/users/\\{alice_id\\}/force-password-reset$")
    public void adminSendsForcePasswordReset() throws Exception {
        String userId = state.getUserId();
        Assertions.assertNotNull(userId, "Alice's user ID must be set");
        ServiceResponse response = AppFactory.getService()
                .adminForcePasswordReset(state.getAdminAccessToken(), userId);
        state.setLastResponse(response);
    }

    @Then("the response body should contain at least one user with {string} equal to {string}")
    public void responseContainsUserWithField(String field, String value) {
        ServiceResponse response = state.getLastResponse();
        Assertions.assertNotNull(response);
        JsonObject body = response.body();
        Assertions.assertNotNull(body);
        JsonArray data = body.getJsonArray("data");
        Assertions.assertNotNull(data, "Expected 'data' array in response");
        boolean found = false;
        for (int i = 0; i < data.size(); i++) {
            JsonObject user = data.getJsonObject(i);
            if (value.equals(user.getString(field))) {
                found = true;
                break;
            }
        }
        Assertions.assertTrue(found,
                "Expected at least one user with '" + field + "' = '" + value + "'");
    }

    @Given("alice's account has been disabled by the admin")
    public void alicesAccountHasBeenDisabledByAdmin() throws Exception {
        String userId = state.getUserId();
        Assertions.assertNotNull(userId);
        AppFactory.getService().adminDisableUser(state.getAdminAccessToken(), userId);
    }

    @Given("alice's account has been disabled")
    public void alicesAccountHasBeenDisabled() throws Exception {
        alicesAccountHasBeenDisabledByAdmin();
    }
}
